extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/health/DataSourceHealthCheck.java

```diff
@@ -0,0 +1,62 @@
+package io.quarkus.agroal.runtime.health;
+
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.spi.Bean;
+import javax.sql.DataSource;
+
+import org.eclipse.microprofile.health.HealthCheck;
+import org.eclipse.microprofile.health.HealthCheckResponse;
+import org.eclipse.microprofile.health.HealthCheckResponseBuilder;
+import org.eclipse.microprofile.health.Readiness;
+
+import io.quarkus.arc.Arc;
+
+@Readiness
+@ApplicationScoped
+public class DataSourceHealthCheck implements HealthCheck {
+    private static final String DEFAULT_DS = "__default__";
+    private Map<String, DataSource> dataSources = new HashMap<>();
+
+    @PostConstruct
+    protected void init() {
+        Set<Bean<?>> beans = Arc.container().beanManager().getBeans(DataSource.class);
+        for (Bean<?> bean : beans) {
+            if (bean.getName() == null) {
+                // this is the default DataSource: retrieve it by type
+                DataSource defaultDs = Arc.container().instance(DataSource.class).get();
+                dataSources.put(DEFAULT_DS, defaultDs);
+            } else {
+                DataSource ds = (DataSource) Arc.container().instance(bean.getName()).get();
+                dataSources.put(bean.getName(), ds);
+            }
+        }
+    }
+
+    @Override
+    public HealthCheckResponse call() {
+        HealthCheckResponseBuilder builder = HealthCheckResponse.named("Database connection(s) health check").up();
+        int cpt = 1;
+        for (Map.Entry<String, DataSource> dataSource : dataSources.entrySet()) {
+            try {
+                boolean valid = dataSource.getValue().getConnection().isValid(0);
+                if (!valid) {
+                    String data = dataSource.getKey() == DEFAULT_DS ? "validation check failed for the default DataSource"
+                            : "validation check failed for DataSource '" + dataSource.getKey() + "'";
+                    builder.down().withData("reason" + cpt++, data);
+                }
+            } catch (SQLException e) {
+                String data = dataSource.getKey() == DEFAULT_DS
+                        ? "Unable to execute the validation check for the default DataSource: "
+                        : "Unable to execute the validation check for DataSource '" + dataSource.getKey() + "': ";
+                builder.down().withData("reason" + cpt++, data + e.getMessage());
```

Comment:
I still think we should use a key other than "reason" to indicate an exception occurred.

Reply:
OK, I propose to use the name of the datasource or 'quarkus-default-ds' for the default one. 
It's very unlinkly that a user will create a named datasource with 'quarkus-default-ds' as its name.